lib/rouge/lexers/armasm.rb

```diff
@@ -0,0 +1,143 @@
+          DCQ DCQU DCW DCWU DN ELIF ELSE END ENDFUNC ENDIF ENDP ENTRY EQU EXPORT
+          EXPORTAS EXTERN FIELD FILL FN FRAME FUNCTION GBLA GBLL GBLS GLOBAL IF
+          IMPORT INFO KEEP LCLA LCLL LCLS LEADR LEAF LTORG MACRO MAP MEND MEXIT
+          NOFP OPT ORG PRESERVE8 PROC QN RELOC REQUIRE REQUIRE8 RLIST RN ROUT
+          SETA SETL SETS SN SPACE STRONG SUBT THUMB THUMBX TTL WEND WHILE
+          \[ \] [|!#*=%&^]
+        )
+      end
+
+      def self.shift_or_condition
+        @shift_or_condition ||= %w(
+          ASR LSL LSR ROR RRX AL CC CS EQ GE GT HI HS LE LO LS LT MI NE PL VC VS
+          asr lsl lsr ror rrx al cc cs eq ge gt hi hs le lo ls lt mi ne pl vc vs
+        )
+      end
+
+      def self.builtin
+        @builtin ||= %w(
+          ARCHITECTURE AREANAME ARMASM_VERSION CODESIZE COMMANDLINE CONFIG CPU
+          ENDIAN FALSE FPIC FPU INPUTFILE INTER LINENUM LINENUMUP LINENUMUPPER
+          OBJASM_VERSION OPT PC PCSTOREOFFSET REENTRANT ROPI RWPI TRUE VAR
+        )
+      end
+
+      def self.operator
+        @operator ||= %w(
+          AND BASE CC CC_ENCODING CHR DEF EOR FATTR FEXEC FLOAD FSIZE INDEX LAND
+          LEFT LEN LEOR LNOT LOR LOWERCASE MOD NOT OR RCONST REVERSE_CC RIGHT ROL
+          ROR SHL SHR STR TARGET_ARCH_[0-9A-Z_]+ TARGET_FEATURE_[0-9A-Z_]+
+          TARGET_FPU_[A-Z_] TARGET_PROFILE_[ARM] UAL UPPERCASE
+        )
+      end
+
+      state :root do
+        rule %r/\n/, Text
+        rule %r/^[ \t]*#[ \t]*(?:(?:#{ArmAsm.preproc_keyword.join('|')})[ \t].*)?\n/, Comment::Preproc
+        rule %r/[ \t]+/, Text, :command
+        rule %r/;.*/, Comment
+        rule %r/\$[a-z_]\w*\.?/i, Name::Namespace # variable substitution or macro argument
+        rule %r/\w+|\|[^|\n]+\|/, Name::Label
+      end
+
+      state :command do
+        rule %r/\n/, Text, :pop!
+        rule %r/[ \t]+/ do |m|
+          token Text
+          goto :args
+        end
+        rule %r/;.*/, Comment, :pop!
+        rule %r/(?:#{ArmAsm.file_directive.join('|')})\b/ do |m|
+          token Keyword
+          goto :filespec
+        end
+        rule %r/(?:#{ArmAsm.general_directive.join('|')})(?=[; \t\n])/, Keyword
+        rule %r/(?:[A-Z][\dA-Z]*|[a-z][\da-z]*)(?:\.[NWnw])?(?:\.[DFIPSUdfipsu]?(?:8|16|32|64)?){,3}\b/, Name::Builtin # rather than attempt to list all opcodes, rely on all-uppercase or all-lowercase rule
+        rule %r/[a-z_]\w*|\|[^|\n]+\|/i, Name::Function # probably a macro name
+        rule %r/\$[a-z]\w*\.?/i, Name::Namespace
+      end
+
+      state :args do
+        rule %r/\n/, Text, :pop!
+        rule %r/[ \t]+/, Text
+        rule %r/;.*/, Comment, :pop!
+        rule %r/(?:#{ArmAsm.shift_or_condition.join('|')})\b/, Name::Builtin
+        rule %r/[a-z_]\w*|\|[^|\n]+\|/i, Name::Variable # various types of symbol
+        rule %r/%[bf]?[at]?\d+(?:[a-z_]\w*)?/i, Name::Label
+        rule %r/(?:&|0x)\h+(?![\hp])/i, Literal::Number::Hex
```

Comment:
Why does the negative lookahead check against `[\hp]`? Shouldn't it just be `p`?

In case it helps, this was the rule as it was in your last commit:

```ruby
rule /(&|0[Xx])[0-9A-Fa-f]+(?![0-9A-FPa-fp])/, Literal::Number::Hex
```

Reply:
I'm no longer sure why I wrote that. I may not have been sure how greedily ruby would match characters otherwise? Possibly a cut-and-paste from the arbitrary base syntax you highlight below? I can't see any reason not to change it for `p`.